test/extensions/filters/http/health_check/config_test.cc

```diff
@@ -100,6 +102,129 @@ TEST(HealthCheckFilterConfig, HealthCheckFilterWithEmptyProto) {
+              Http::FilterHeadersStatus::Continue);
+  }
+}
+
+TEST(HealthCheckFilterConfig, HealthCheckFilterHeaderMatch) {
+  envoy::config::filter::http::health_check::v2::HealthCheck config;
+
+  config.mutable_pass_through_mode()->set_value(false);
+
+  envoy::api::v2::route::HeaderMatcher& xheader = *config.add_headers();
+  xheader.set_name("x-healthcheck");
+
+  envoy::api::v2::route::HeaderMatcher& yheader = *config.add_headers();
+  yheader.set_name("y-healthcheck");
+  yheader.set_value("foo");
+
+  Http::TestHeaderMapImpl headers{{"x-healthcheck", "arbitrary_value"}, {"y-healthcheck", "foo"}};
+
+  TestHealthCheckHeaderMatch(config, headers, true);
+}
+
+TEST(HealthCheckFilterConfig, HealthCheckFilterHeaderMatchWrongValue) {
+  envoy::config::filter::http::health_check::v2::HealthCheck config;
+
+  config.mutable_pass_through_mode()->set_value(false);
+
+  envoy::api::v2::route::HeaderMatcher& xheader = *config.add_headers();
+  xheader.set_name("x-healthcheck");
+
+  envoy::api::v2::route::HeaderMatcher& yheader = *config.add_headers();
+  yheader.set_name("y-healthcheck");
+  yheader.set_value("foo");
+
+  Http::TestHeaderMapImpl headers{{"x-healthcheck", "arbitrary_value"}, {"y-healthcheck", "bar"}};
+
+  TestHealthCheckHeaderMatch(config, headers, false);
+}
+
+TEST(HealthCheckFilterConfig, HealthCheckFilterHeaderMatchMissingHeader) {
+  envoy::config::filter::http::health_check::v2::HealthCheck config;
+
+  config.mutable_pass_through_mode()->set_value(false);
+
+  envoy::api::v2::route::HeaderMatcher& xheader = *config.add_headers();
+  xheader.set_name("x-healthcheck");
+
+  envoy::api::v2::route::HeaderMatcher& yheader = *config.add_headers();
+  yheader.set_name("y-healthcheck");
+  yheader.set_value("foo");
+
+  Http::TestHeaderMapImpl headers{{"y-healthcheck", "foo"}};
+
+  TestHealthCheckHeaderMatch(config, headers, false);
+}
+
+TEST(HealthCheckFilterConfig, HealthCheckFilterEndpoint) {
+  envoy::config::filter::http::health_check::v2::HealthCheck config;
+
+  config.mutable_pass_through_mode()->set_value(false);
+
+  config.set_endpoint("foo");
+
+  envoy::api::v2::route::HeaderMatcher& header = *config.add_headers();
+  header.set_name(Http::Headers::get().Path.get());
+  header.set_value("bar");
+
+  Http::TestHeaderMapImpl headers{{Http::Headers::get().Path.get(), "foo"}};
+
+  TestHealthCheckHeaderMatch(config, headers, true);
+}
+
+TEST(HealthCheckFilterConfig, HealthCheckFilterEndpointOverride) {
+  envoy::config::filter::http::health_check::v2::HealthCheck config;
+
+  config.mutable_pass_through_mode()->set_value(false);
+
+  config.set_endpoint("foo");
+
+  envoy::api::v2::route::HeaderMatcher& header = *config.add_headers();
+  header.set_name(Http::Headers::get().Path.get());
+  header.set_value("bar");
+
+  Http::TestHeaderMapImpl headers{{Http::Headers::get().Path.get(), "bar"}};
+
+  TestHealthCheckHeaderMatch(config, headers, false);
+}
+
```

Comment:
might be worth it to add duplicate header matcher test, what is the behavior when the same header name is specified more than once?